opennebula/resource_opennebula_virtual_machine.go

```diff
@@ -1504,6 +1535,27 @@ func generateVm(d *schema.ResourceData, tplContext *dyn.Vector) (string, error)
 }
 
 func resourceVMCustomizeDiff(diff *schema.ResourceDiff, v interface{}) error {
+
+	onChange := diff.Get("on_disk_change").(string)
+	oldDisk, newDisk := diff.GetChange("disk")
+
+	oldDiskList := oldDisk.([]interface{})
+	newDiskList := newDisk.([]interface{})
+
+	for i, oi := range oldDiskList {
+		oiB := oi.(map[string]interface{})
+		niB := newDiskList[i].(map[string]interface{})
+
+		log.Printf("[DEBUG] old disk: %s\n", oiB["image_id"])
+		log.Printf("[DEBUG] new disk: %s\n", niB["image_id"])
+
+		if oiB["image_id"].(int) != niB["image_id"].(int) {
```

Comment:
The full behaviour around disk changes wasn't defined, among other we could have some corner cases. Are they acceptable for everyone (feel free to discuss) ?

For instance: 
- Even if the user add a new disk (other than the OS disk like described in the issue), this will trigger the behavior (recreate/restart).
- If an user swap two disk image IDs other than the OS disk (and potentially update some other fields  of the disks): from A B C to A C B
  we'll have almost have the same disk list, but this change will make the provider recreate/restart the VM

Reply:
The original intention was to modify behavior for when an attached disk was changed, not added or removed. If you have other behavioral suggestions, I will happily implement them too.

Reply:
I'm ok with the behavior, however it remains some work. Below I'll try to be clear and as accurate as I can, so sorry for the bunch of text.

I consider a disk as "changed" when some attributes changed, so compare only the image_id of the disks is not sufficient.
If other attributes than image_id changed (target, driver...) this will trigger a detachment followed by an attachment of the disk (AFAIK there is no update-in-place method for disk in OpenNebula, except resize method).
So, we want to find the disks that have the same image_id with some attributes changed to know if we should trigger the on_disk_change behavior.
For now I'll ignore the case when an image would be attached twice to the same VM (two disk with same image_id, introduced by #125).

They is several ways to implement this: from scratch with a double loop to compare the list of disks and find disks changed, or inspired on how I managed disk update, or a better idea if you have one.

Here is how I managed disk update:
I added an helper to retrieve the lists of disks to attach/detach:
https://github.com/OpenNebula/terraform-provider-opennebula/search?q=diffListConfig
As you can see, I compare some sets of disks based on attributes image_id, target, driver (changes on other attributes would be ignored).
The helper returns us toAttach and toDetach lists, but they include more disks that we want: toAttach contains new disks to attach (new image_id introduced), and toDetach contains disk that we only want to detach.
At least, there is some additional filtering to do on these list to determine if there is some disks changed.

I hope this help, it's nice to have other externals contributors so feel free to drop some comments if I can help more

Reply:
@danhigham about this comment ?

Reply:
Apologies for the delay on this, I will work on getting these changes in this week.